plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinHighLevelFunctionParameterInfoHandler.kt

```diff
@@ -0,0 +1,454 @@
+
+    override fun getActualParametersRBraceType(): KtSingleValueToken = KtTokens.RPAR
+
+    override fun getArgumentListAllowedParentClasses() = setOf(KtCallElement::class.java)
+}
+
+class KotlinHighLevelLambdaParameterInfoHandler :
+    KotlinHighLevelParameterInfoWithCallHandlerBase<KtLambdaArgument, KtLambdaArgument>(KtLambdaArgument::class, KtLambdaArgument::class) {
+
+    override fun getActualParameters(lambdaArgument: KtLambdaArgument) = arrayOf(lambdaArgument)
+
+    override fun getActualParametersRBraceType(): KtSingleValueToken = KtTokens.RBRACE
+
+    override fun getArgumentListAllowedParentClasses() = setOf(KtLambdaArgument::class.java)
+
+    override fun getParameterIndex(context: UpdateParameterInfoContext, argumentList: KtLambdaArgument): Int {
+        val size = (argumentList.parent as? KtCallElement)?.valueArguments?.size ?: 1
+        return size - 1
+    }
+}
+
+class KotlinHighLevelArrayAccessParameterInfoHandler :
+    KotlinHighLevelParameterInfoWithCallHandlerBase<KtContainerNode, KtExpression>(KtContainerNode::class, KtExpression::class) {
+
+    override fun getArgumentListAllowedParentClasses() = setOf(KtArrayAccessExpression::class.java)
+
+    override fun getActualParameters(containerNode: KtContainerNode): Array<out KtExpression> =
+        containerNode.allChildren.filterIsInstance<KtExpression>().toList().toTypedArray()
+
+    override fun getActualParametersRBraceType(): KtSingleValueToken = KtTokens.RBRACKET
+}
+
+// TODO: Overall structure is similar to FE 1.0 version (KotlinParameterInfoWithCallHandlerBase). Extract common logic to base class.
+abstract class KotlinHighLevelParameterInfoWithCallHandlerBase<TArgumentList : KtElement, TArgument : KtElement>(
+    private val argumentListClass: KClass<TArgumentList>,
+    private val argumentClass: KClass<TArgument>
+) : ParameterInfoHandlerWithTabActionSupport<TArgumentList, KotlinHighLevelParameterInfoWithCallHandlerBase.CandidateInfo, TArgument> {
+
+    companion object {
+        @JvmField
+        val GREEN_BACKGROUND: Color = JBColor(Color(231, 254, 234), Gray._100)
+
+        val STOP_SEARCH_CLASSES: Set<Class<out KtElement>> = setOf(
+            KtNamedFunction::class.java,
+            KtVariableDeclaration::class.java,
+            KtValueArgumentList::class.java,
+            KtLambdaArgument::class.java,
+            KtContainerNode::class.java,
+            KtTypeArgumentList::class.java
+        )
+    }
+
+    override fun getActualParameterDelimiterType(): KtSingleValueToken = KtTokens.COMMA
+
+    override fun getArgListStopSearchClasses(): Set<Class<out KtElement>> = STOP_SEARCH_CLASSES
+
+    override fun getArgumentListClass() = argumentListClass.java
+
+    override fun showParameterInfo(element: TArgumentList, context: CreateParameterInfoContext) {
+        context.showHint(element, element.textRange.startOffset, this)
+    }
+
```

Comment:
This code is the same as in FE1.0 version. Maybe it can be extracted to a base class?

Reply:
I have a TODO to come up with a common base class. The structure is generally similar, but I wanted to do it separately.